scylla-cdc/src/log_reader.rs

```diff
@@ -350,18 +409,45 @@ impl CDCLogReaderBuilder {
             tokio::sync::watch::channel(end_timestamp);
         let readers = vec![];
 
+        let mut checkpoint_saver = self.checkpoint_saver;
+        if (self.should_save_progress || self.should_load_progress) && checkpoint_saver.is_none() {
+            checkpoint_saver = Some(Arc::new(
+                CDCDefaultCheckpointSaver::new(
+                    &session,
+                    &keyspace,
+                    Some(&*(table_name.clone() + "_checkpoints")),
+                )
+                .await?,
+            ));
+        }
+        let mut start_timestamp = self.start_timestamp;
+        if self.should_load_progress {
+            if let Some(generation) = checkpoint_saver
+                .as_ref()
+                .unwrap()
+                .load_last_generation()
+                .await?
+            {
+                start_timestamp = generation.timestamp;
```

Comment:
I think there is a potential problem. Consider the following situation:

1. You created a cluster a week ago and didn't make any topology changes since,
2. You start the application with progress saving turned on and tell it to start reading from the `now` timestamp,
3. The application saves the generation but crashes immediately after that, without saving information about progress of any of the streams,
4. You restart the application, it loads the generation timestamp (which is set to a week ago), the stream readers see that no progress was saved for them and they start reading from the beginning of the generation.

I think we should store both the start timestamp and the generation timestamp in the progress table.

Reply:
Okay, I understand why `start_timestamp = generation.timestamp;` is a bad idea, but I think setting `start_timestamp` to `max(last_generation, start_timestamp` should be enough instead of storing the start timestamp. Am I missing something here?

Reply:
OK, we can use `max(last_generation, start_timestamp)`.